extractor.go

```diff
@@ -316,3 +322,16 @@ func match(r io.Reader) (io.Reader, types.Type, error) {
 
 	return r, typ, err
 }
+
+// safeJoin performs a filepath.Join of 'parent' and 'subdir' but returns an error
+// if the resulting path points outside of 'parent'.
+func safeJoin(parent, subdir string) (string, error) {
+	res := filepath.Join(parent, subdir)
+	if !strings.HasSuffix(parent, string(os.PathSeparator)) {
+		parent += string(os.PathSeparator)
+	}
+	if !strings.HasPrefix(filepath.Clean(res), parent) {
```

Comment:
`Join` clean the path after joining. See [the docs](https://godoc.org/path/filepath#Join). As such beware of cleaning it multiple times. 

Is also possible that `parent` (that is not being `Clean`ed) will never prefix-match `res` (that is being `Clean`ed). Something worth checking out.

This is the reference code from Snyk ZipSlip:
```go
if !strings.HasPrefix(destpath, filepath.Clean(destination) + string(os.PathSeparator)) {
    return fmt.Errorf("%s: illegal file path", filePath)
}
```

Reply:
> Join clean the path after joining. See the docs. As such beware of cleaning it multiple times.

Good point! will push a fix for that.

Reply:
> This is the reference code from Snyk ZipSlip:

the full function from Snyk is:

```go
func sanitizeExtractPath(filePath string, destination string) error {
    destpath := filepath.Join(destination, filePath)
    if !strings.HasPrefix(destpath, filepath.Clean(destination) + string(os.PathSeparator)) {
        return fmt.Errorf("%s: illegal file path", filePath)
    }
    return nil
}
```

I've found that the function gives the wrong output in this particular case:

* `sanitizeExtractPath("legal/path", "/")`

it will return an error but the path is valid, this is because I think they assume that a "cleaned" path will always miss the path separator `/` at the end of the string but this is not true for the root path `"/"` that once cleaned is still `"/"` so the `filepath.Clean(destination) + string(os.PathSeparator)` will end up to be `"//"` that can not be the prefix of any joined path `"/legal/path"`.